src/PrintCore/PrintCore.cs

```diff
@@ -0,0 +1,775 @@
+				if (code != PMStatusCode.Ok)
+					return null;
+				return CFString.FetchString (s);
+			}
+		}
+
+		public double Width {
+			get {
+				double s; 
+				var code = PMPaperGetWidth (handle, out s);
+				if (code != PMStatusCode.Ok)
+					return 0;
+				return s;
+			}
+		}
+			
+		public double Height {
+			get {
+				double s; 
+				var code = PMPaperGetHeight (handle, out s);
+				if (code != PMStatusCode.Ok)
+					return 0;
+				return s;
+			}
+		}
+
+		public PMPaperMargins? Margins {
+			get {
+				PMPaperMargins margins;
+				var code = PMPaperGetMargins (handle, out margins);
+				if (code != PMStatusCode.Ok)
+					return null;
+				return margins;
+			}
+		}
+
+		public string GetLocalizedName (PMPrinter printer)
+		{
+			if (printer == null)
+				throw new ArgumentNullException (nameof (printer));
+			IntPtr name;
+			var code = PMPaperCreateLocalizedName (handle, printer.handle, out name);
+			if (code != PMStatusCode.Ok)
+				return null;
+			return CFString.FetchString (name);
+		}
+		
+	}
+
+	public class PMPrinter : PMPrintCoreBase {
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMCreateGenericPrinter (out IntPtr session);
+		
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static IntPtr PMPrinterCreateFromPrinterID (IntPtr id);
+
+		internal PMPrinter (IntPtr handle, bool owns) : base (handle, owns) {}
+		public PMPrinter ()
+		{
+			IntPtr value;
+			var code = PMCreateGenericPrinter (out value);
+			if (code == PMStatusCode.Ok)
+				handle = value;
+			else
+				throw new PMPrintException (code);
+		}
+
+		public PMPrinter (string printerId)
+		{
+			using (var idf = new CFString (printerId)){
+				var value = PMPrinterCreateFromPrinterID (idf.Handle);
+				if (value == IntPtr.Zero)
+					throw new PMPrintException (PMStatusCode.InvalidPrinter);
+					
+				handle = value;
+			}
+		}
+		
+		public static PMStatusCode TryCreate (out PMPrinter printer)
+		{
+			IntPtr value;
+			var code = PMCreateGenericPrinter (out value);
+			if (code == PMStatusCode.Ok){
+				printer = new PMPrinter (value, owns: true);
+				return PMStatusCode.Ok;
+			}
+			printer = null;
+			return code;
+		}
+
+		public static PMPrinter TryCreate (string printerId)
+		{
+			using (var idf = new CFString (printerId)){
+				var h = PMPrinterCreateFromPrinterID (idf.Handle);
+				if (h == IntPtr.Zero)
+					return null;
+				return new PMPrinter (h, owns: true);
+			}
+		}
+
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static IntPtr PMPrinterGetName (IntPtr handle);
+		public string Name => CFString.FetchString (PMPrinterGetName (handle));
+
+		[DllImport (Constants.PrintCoreLibrary)]
+		extern static PMStatusCode PMPrinterCopyDeviceURI (IntPtr handle, out IntPtr url);
+
+		public PMStatusCode TryGetDeviceUrl (out NSUrl url)
+		{
+			IntPtr urlH;
+			
+			var code = PMPrinterCopyDeviceURI (handle, out urlH);
+			if (code != PMStatusCode.Ok){
+				url = null;
+				return code;
+			}
+			url = new NSUrl (urlH);
```

Comment:
This leaks the NSUrl, `PMPrinterCopyDeviceURI` gives you ownership.

You can use `Runtime.GetNSObject (urlH, true)` instead which will do the right thing.